core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,666 @@
+              result
+          }
+        loop(h :: t.toList)
+      case NormalPattern.PositionalStruct(maybeIdx, items) =>
+        // The type in question is not the outer dt, but the type associated
+        // with this current constructor
+        val itemFns = items.map(maybeBind(_))
+
+        def processArgs(as: List[NormalExpression], acc: PatternEnv): PatternMatch[PatternEnv] = {
+          // manually write out foldM hoping for performance improvements
+          @annotation.tailrec
+          def loop(vs: List[NormalExpression], fns: List[(NormalExpression, PatternEnv) => PatternMatch[PatternEnv]], env: (PatternEnv, PatternMatch[PatternEnv])): PatternMatch[PatternEnv] =
+            vs match {
+              case Nil => env._2
+              case vh :: vt =>
+                fns match {
+                  case fh :: ft =>
+                    (fh(vh, env._1), env._2) match {
+                      case (_, NoMatch) => NoMatch
+                      case (NoMatch, _) => NoMatch
+                      case (Matches(env1), Matches(_)) => loop(vt, ft, (env1, Matches(env1)))
+                      case (Matches(env1), NotProvable) => loop(vt, ft, (env1, NotProvable))
+                      case (NotProvable, _) => loop(vt, ft, (env._1, NotProvable))
+                    }
+                  case Nil => env._2 // mismatch in size, shouldn't happen statically
+                }
+            }
+          loop(as, itemFns, (acc, Matches(acc)))
+        }
+
+        maybeIdx match {
+          case None => 
+            // this is a struct, which means we expect it
+            { (arg: NormalExpression, acc: PatternEnv) =>
+              arg match {
+                case NormalExpression.Struct(_, args) =>
+                  processArgs(args, acc)
+                case _ =>
+                  NotProvable
+              }
+            } 
+
+          case Some(idx) =>
+            // we don't check if idx < 0, because if we compiled, it can't be
+            val result = { (arg: NormalExpression, acc: PatternEnv) =>
+              arg match {
+                case NormalExpression.Struct(enumId, args) =>
+                  if (enumId == idx) processArgs(args, acc)
+                  else NoMatch
+                case _ =>
+                  NotProvable
+              }
+            }
+            result
+        }
+  }
+
+  def findMatch(m: NormalExpression.Match) =
+    m.branches.collectFirst(Function.unlift( { case (pat, result) =>
+      maybeBind(pat).apply(m.arg, Map()) match {
+        case Matches(env) => Some(Some((pat, env, result)))
+        case NotProvable => Some(None)
+        case NoMatch => None
+      }
+    })).get // Totallity of matches should ensure this will always find something unless something has gone terribly wrong
+
+  def solveMatch(env: PatternEnv, result: NormalExpression) =
+    ((env.size - 1) to 0 by -1).toList.map(env.get(_).get) // If this exceptions then somehow we didn't get enough names in the env
```

Comment:
why the .toList? seems not needed. I think `.iterator` would be slightly better for perf so we wouldn't reallocate a copy in the map.

Reply:
huh, yeah. That's clearly right. Don't know what I was thinking